### PR TITLE
Eloi/autosuspend optimization

### LINKED ISSUE #32: Add utils sql file
There are several views, tables or udfs that we could create in a general utils sql file that can also live in the `procedures` directory. This could include, for instance, UDTFs that return the `show objects` commands as a table usable in queries.

@wiesnseb @DLampriadis should we just call it `utils.sql`? Anything else that you want to add?

### PR DESCRIPTION
**Pull request by EloiSanchez**
## Summary
Significant improvement on the auto suspend check.
- Added cost calculation and prediction when increasing the warehouse auto suspend time
- Created the utils.sql file with the procedure that will generate useful tables for the health checks

## Does this PR refer to any issue?
Fixes #32 but keeps open #13, since there are things to do. 

### Files
- auto_suspend.sql
- auto_suspend.py
- utils.sql

### Procedure
- auto_suspend()
- create_tables_info() in utils.sql
- create_warehouse_info() in utils.sql

### Tables
- auto_suspend_results
- auto_suspend_intermediate
- warehouse_info

### Functions
- auto_suspend_compute_cost(new_auto_suspend int)

## I made sure that...
- [x] The files are named `{health_check_name}.sql/py`, are saved in the proper directory and are well documented
- [x] The procedure is named `{health_check_name}` and is stored in the schema `health_checks.public`
- [x] The tables created are named `{health_check_name}[_extra_info]_[results or intermediate]` and are stored in the schema `health_checks.public`
- [x] I run the pre-commit hooks in all my commits

## Does my code code need to be tested?
Yes

## The reviewer should check
1. Check the autosuspend procedure
2. Run the Python script and check the plots
3. Check the utils file
4. Make sure that the calculations for the cost predictions are correct and make sense

### CODE REVIEW

**Comment by EloiSanchez** [comment]
New requisites:

- Instead of using the objects accessed, use the suspended periods for the warehouses
- Calculate the cost taking into account warehouse size
- Calculate the cost taking into account snowpark-optimized warehouses
- Calculate the amount of times a warehouse would be started by using different autosuspend times.
- Create a new graph that shows all the information and graphs for a single warehouse
- Investigate how warehouse cache works in multiprocessing warehouses
- Calculate the cost taking into account multiprocessing (it depends on how the result of the research on the previous point goes, maybe it is not needed)

File: snowflake-health-checks/src/libraries/procedures/auto_suspend.sql [hunk 1]
```diff
@@ -97,16 +158,272 @@ begin
...
+        ),
+
+        suspended_times as (
+            select
+                event_history.warehouse_name,
+                (event_history.period_time + wh_info.auto_suspend) / 60 as inactive_time
+            from event_history
+            left join health_checks.public.warehouse_info as wh_info
+                on event_history.warehouse_name = wh_info.name
+        ),
```
Comment (EloiSanchez) [inline snowflake-health-checks/src/libraries/procedures/auto_suspend.sql:216]: This has to be deleted. It is not used

Comment (EloiSanchez) [reply snowflake-health-checks/src/libraries/procedures/auto_suspend.sql]: done in [71e1516](https://github.com/EloiSanchez/snowflake-health-check/pull/38/commits/71e151682827e029478c932292b24e30206ce5dd)

**Comment by EloiSanchez** [comment]
Done:

- Instead of using the objects accessed, use the suspended periods for the warehouses
- Calculate the amount of times a warehouse would be started by using different autosuspend times.
- Create a new graph that shows all the information and graphs for a single warehouse
- Investigate how warehouse cache works in multiprocessing warehouses

**Summary**
Two methods exist in the plot object, one to create the overview for all the warehouses and one with the detail for a single warehouse.

## Does this PR refer to any issue?
Fixes #32 but keeps open #13, since there are things to do.

## Objects created or modified
### Files
- auto_suspend.sql
- auto_suspend.py
- utils.sql
- connection.py
- Home.py
- Warehouse overview.py
- requirements.txt

### Procedure
- auto_suspend()
- create_tables_info() in utils.sql
- create_warehouse_info() in utils.sql

### Tables
- auto_suspend_results
- auto_suspend_acc_objs_intermediate
- auto_suspend_time_suspended_intermediate
- auto_suspend_histogram_data_intermediate
- warehouse_info

### Functions
- auto_suspend_predictions(new_auto_suspend int)

## I made sure that...
- [x] The files are named `{health_check_name}.sql/py`, are saved in the proper directory and are well documented
- [x] The procedure is named `{health_check_name}` and is stored in the schema `health_checks.public`
- [x] The tables created are named `{health_check_name}[_extra_info]_[results or intermediate]` and are stored in the schema `health_checks.public`
- [x] I run the pre-commit hooks in all my commits

## Does my code code need to be tested?
Yes

## The reviewer should check
1. Check the autosuspend procedure
2. Run the Python script and check the plots
3. Check the utils file
4. Make sure that the calculations for the cost predictions are correct and make sense
5. Run streamlit app and see the plots in the dashboard

**Review by DLampriadis — APPROVED** [review]
Everything works fine.

**Pull request merged**